### PR TITLE
Implement System.ComponentModel.VersionConverter

### PR DESCRIPTION
**Pull request by 0xced**
VersionConverter is a new System.ComponentModel.TypeConverter subclass that handle conversions between string and System.Version.

Edit @stephentoub 6/12/2018: https://github.com/dotnet/corefx/issues/28594

### CODE REVIEW

**Comment by karelz** [comment]
@0xced thanks for your PR. You are adding new public API, which needs to go through [API review](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md) first.
Can you please file an issue with API description, motivation, etc. and work with area owners (@maryamariyan @safern) to agree on that first? Thanks!

**Comment by danmosemsft** [comment]
@karelz @wfurt for System.Net.HttpListener.Tests and System.Net.NameResolution.Functional.Tests can we move them out of inner loop? We don't want any flaky tests in CI.

**Comment by danmosemsft** [comment]
For the other issues I would do git clean -fdx then try again.

To be clear though this API needs approval from the API review board. They might not convene for a couple weeks, and might reject it or want changes...

**Comment by danmosemsft** [comment]
@0xced the API was approved, can you make any necessary changes to match the approved shape, and get tests green?

@dotnet-bot test this please

**Comment by 0xced** [comment]
I keep rebasing on master but `NETFX x86 Release`, `UWP CoreCLR x64 Debug` and `UWP NETNative x86 Release` keep failing while all other platforms are fine.

Does anybody has an idea what’s happening with those three platforms?

**Comment by stephentoub** [comment]
> Does anybody has an idea what’s happening with those three platforms?

The new type you've added only exists in .NET Core; it doesn't exist in .NET Framework or UWP.  So the tests when built for .NET Framework or UWP are failing to compile.  You need to ensure that your new tests are only built when targeting .NET Core, e.g. with a condition in the project file like `Condition="'$(TargetGroup)' == 'netcoreapp'"`.

File: src/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs]: Why did you undo my change and switch to using an ifdef?  We try hard to avoid such ifdefs, and instead prefer approaches like I suggested.

Comment (0xced) [reply src/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs]: Sorry, I did not see you did a change, I rebased on master and force pushed. Adding `Condition="'$(TargetGroup)' == 'netcoreapp'"` was not enough because of this occurence of `VersionConverter` in `TypeDescriptorTests.cs`.

What was your approach to avoid the conditional compilation?

Comment (stephentoub) [reply src/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs]: Not to avoid conditional compilation... to avoid the ifdef.  And also to use theories rather than iterating through an enumerable, as theories are generally the approach we prefer in this case.
https://github.com/stephentoub/corefx/commit/7d695a634336fc925a3bc5573402faae38a4ee54

Comment (0xced) [reply src/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs]: OK thanks. I reverted my `#if NETCORE` fix, cherry-picked yours, rebased on master and force-pushed.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs]: It looks like other ConvertFrom methods trim the input string before trying to parse it.  Could you check what the behavior is there, and then potentially adapt this accordingly to match if it should trim?  Same question for IsValid

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs]: Using `Version.Parse` would be better; the ctor just calls `Version.Parse` and copies data out of the result into `this`, so going straight to `Version.Parse` saves an allocation.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs]: The above might throw an ArgumentException.  Is that desired?   Should it instead be a FormatException?  What do the docs for ConvertFrom state?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs]: Other TypeConverters appear to delegate to the base implementation (which in turn calls this).  We should follow the same pattern unless there's a really good reason not to.

**Comment by stephentoub** [comment]
cc: @safern, @maryamariyan

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs
Comment (0xced) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs]: The default behavior of Version.Parse is to ignore all whitespaces, so this is actually already taken care of. I added some test data in cfc4ea107f2807a6d84d41cf15a9c5ab8d1daa73 to ensure correct behavior.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs
Comment (0xced) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs]: Done in d8d780c969fe696be8f764a41ae9edd9d6ecb6bf.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs
Comment (0xced) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs]: This was discussed in https://github.com/dotnet/corefx/issues/28594#issuecomment-377323389 and @safern agreed on
> let's keep it simple and just let the constructor handle it itself

But the doc from [ConvertFrom](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.typeconverter.convertfrom?view=netcore-2.1#System_ComponentModel_TypeConverter_ConvertFrom_System_ComponentModel_ITypeDescriptorContext_System_Globalization_CultureInfo_System_Object_) says
> Exceptions
> [NotSupportedException](https://docs.microsoft.com/en-us/dotnet/api/system.notsupportedexception?view=netcore-2.1)
> The conversion cannot be performed.

And `NotSupportedException` is used in the base `TypeConverter` class when the value is not of a supported type (`string` or `Version`).

I'm not sure what’s the best solution here.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs
Comment (0xced) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs]: You're right, there is no good reason, I addressed this in a5ad73cc1ebfbec15ca1d5cdc007cf72654344fe.

The origin of throwing directly is that I used `UriTypeConverter` as a model to implement `VersionConverter` and `UriTypeConverter` throws directly without calling the base class implementation.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs
Comment (stephentoub) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs]: Ok, thanks.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs
Comment (stephentoub) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs]: @safern, I'll leave it up to you.  Just seems strange to me that we'd differ from what's doc'd here.  Are the docs wrong / are the other type converters already inconsistent and throw exceptions like ArgumentException?

cc: @danmosemsft

**Comment by 0xced** [comment]
I just rebased on master and 🎉 **all the tests are finally green**. ✅

**Review by safern — COMMENTED** [review]
LGTM, thanks @0xced

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs
Comment (safern) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs]: I talked to @stephentoub offline since I missed this discussion, I think it is a fair point to try to keep consistency across newly added converters and try to fit what is best. As a conclusion after looking at other converters this are the points I would like to mention:

- The docs are referring to the base class as @0xced states which for me is correct as that should be thrown only when the type converting from is not a supported type to convert from. In this case not `string` or `Version` which we're doing in all other converters. However it might be worth mentioning that in overridden methods we could throw `ArgumentException` or `FormatException` 

- The other exceptions thrown are `ArgumentException` and `FormatException` in all other converters, but they are not leaving it to the Parser to throw the exception, they are catching the exception and throwing a custom one. 

So I think the right thing to do here is to catch the exception and throw a `FormatException` as I suggested earlier in the issue. The reason why I believe that is the right outcome is because, that way the exception is clear that when trying to convert from a supported type we couldn't convert it because the format of the value was incorrect. That is what for me is the clearest approach and that way we would maintain a "sort of consistency" where we only throw ArgumentException, FormatException or NotSupportedException. 

Version.Parse throws this exceptions

ArgumentNullException | input is null.
-- | --
ArgumentException | input has fewer than two or more than four version components.
ArgumentOutOfRangeException | At least one component in input is less than zero.
FormatException | At least one component in input is not an integer.
OverflowException | At least one component in input represents a number that is greater than Int32.MaxValue.

So I wouldn't mind catching everything but FormatException and converting that to ArgumentException if we think that would be a best fit.

**Comment by karelz** [comment]
@safern why did you mark it as NO MERGE?
This is fairly old PR, how can we push it forward?

**Comment by safern** [comment]
I added NO MERGE because I'm waiting for response on my comment above: https://github.com/dotnet/corefx/pull/28516#discussion_r202467189

**Comment by danmosemsft** [comment]
@0xced could you respond to @safern feedback? seem that is all that is left.

**Comment by 0xced** [comment]
I'm currently away from computer, I will reply in a few weeks.

**Comment by danmosemsft** [comment]
@0xced sure no problem. I'll let this PR sit then.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs
Comment (0xced) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/VersionConverter.cs]: I agree that throwing `FormatException` is better. Unfortunately, I only looked at `GuidConverter` and `UriTypeConverter` which both don't catch any exceptions thrown by the constructors.

It is now addressed in 004ef11348398f7dbdd65405d6fe1a9e65889076 and rebased on master.

**Review by safern — APPROVED** [review]
LGTM.

**Comment by safern** [comment]
Thanks @0xced

**Comment by 0xced** [comment]
Do you know when it will land in .NET Core? I just checked in the recently released version 2.2 and `System.ComponentModel.VersionConverter` does not exist.

**Comment by karelz** [comment]
@0xced it is in 3.0 milestone. Try 3.0 Preview1 - it went out today as well.

**Comment by 0xced** [comment]
Excellent! I didn't see it was in a milestone, thanks.

**Comment by hughbe** [comment]
Couple of things that I would like to point out that are inconsistent between `UriTypeConverter/VersionConverter` and the other type converters in .NET Core:
- They throw `ArgumentNullException` for `CanConvertFrom(null)` whereas .NET Framework returns `false`. I suggest fixing this to be consistent with other type converters.
- They return `false` for `CanConvertFrom(typeof(InstanceDescriptor))` because they don't call their base class `CanConvertFrom` method. See https://github.com/dotnet/corefx/pull/35760. I suggest fixing this
- They don't support `InstanceDescriptor` conversion - should this be added to `VersionConverter`?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs
Comment (niemyjski) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs:129]: Sure would be nice if every type listed here had a comment on the class itself or in docs someplace that stated it had a type converter. The first place I looked when serialization broke for me on the upgrade to 3.1 from 2.2 was on the Version class and I didn't see it. I searched for usages on GitHub and it's parser must have problems with `[typeof(Version)]` as I searched for `typeof(Version)` and `typeof(System.Version)` with no luck. https://github.com/JamesNK/Newtonsoft.Json/issues/2243

**Pull request merged**
